src/pypdfium2/_helpers/page.py

```diff
@@ -459,6 +459,32 @@ def render(
         if (color_scheme is not None) and fill_to_stroke:
             flags |= pdfium_c.FPDF_CONVERT_FILL_TO_STROKE
         
+        # Check for non-embedded fonts and warn if found
+        missing_fonts = _check_missing_fonts(self)
+        if missing_fonts:
+            warnings = []
+            for font_name in sorted(missing_fonts.keys()):
+                text_samples = missing_fonts[font_name]
+                if text_samples:
+                    # Limit number of text samples per font to keep warning readable
+                    max_samples = 3
+                    if len(text_samples) > max_samples:
+                        shown = text_samples[:max_samples]
+                        remaining = len(text_samples) - max_samples
+                        text_display = ", ".join(f'"{t}"' for t in shown) + f" (and {remaining} more)"
+                    else:
+                        text_display = ", ".join(f'"{t}"' for t in text_samples)
+                    warnings.append(f"{font_name} (used in: {text_display})")
+                else:
+                    # Fallback if we couldn't extract any text samples
+                    warnings.append(f"{font_name} (text content unavailable)")
+            
+            font_list = "; ".join(warnings)
+            logger.warning(
+                f"Page contains non-embedded fonts that may not render correctly if not available on the system: {font_list}. "
+                f"Text using these fonts may be missing from the rendered image."
+            )
```

Comment:
Thank you for providing this code, I was able to take a quick look after all.

Could you factor out this implementation into a separate method to avoid lengthening the main `page.render()` method? Also, it may be better to make this optional. Opt-in, I think, as this potentially involves many API calls.

Reply:
Or maybe just have it as a standalone public API, which the caller can invoke before `.render()` ?